### ISSUE TITLE
the formual is wrong ?

### ISSUE THREAD

**Issue by hongyesuifeng** [jhwjhw0123/Xgboost-With-Imbalance-And-Focal-Loss#1]
![focal_loss](https://user-images.githubusercontent.com/22854275/51952349-21e4ea00-2473-11e9-8aa0-47daf5dad25a.PNG)
is the y and y_pred reversal？

**Comment by jhwjhw0123** [comment]
The notation \hat{y} is the ground-truth label (guess is what you mean 'y'), and the notation y here is the prediction.

**Comment by shaojunchao** [comment]
It seems that the formula of the 1st-order derivative(the second term) is still wrong.

**Comment by jhwjhw0123** [comment]
> It seems that the formula of the 1st-order derivative(the second term) is still wrong.

You mean the (y+\hat{y}-1)^{\gamma} term?

**Comment by shaojunchao** [comment]
> > It seems that the formula of the 1st-order derivative(the second term) is still wrong.
> 
> You mean the (y+\hat{y}-1)^{\gamma} term?

I mean the last one, (\hat{y}-y)^{\gamma+1} term

**Comment by jhwjhw0123** [comment]
> > > It seems that the formula of the 1st-order derivative(the second term) is still wrong.
> > 
> > 
> > You mean the (y+\hat{y}-1)^{\gamma} term?
> 
> I mean the last one, (\hat{y}-y)^{\gamma+1} term

This part is unlikely to be wrong as if you plug in \gamma = 0 it will recover ordinary cross-entropy derivative. Nevertheless, do you have a result you believe to be correct?

**Comment by shaojunchao** [comment]
> > > > It seems that the formula of the 1st-order derivative(the second term) is still wrong.
> > > 
> > > 
> > > You mean the (y+\hat{y}-1)^{\gamma} term?
> > 
> > 
> > I mean the last one, (\hat{y}-y)^{\gamma+1} term
> 
> This part is unlikely to be wrong as if you plug in \gamma = 0 it will recover ordinary cross-entropy derivative. Nevertheless, do you have a result you believe to be correct?

When \hat{y} is 0, the gradient excluding the log term should be (y)^{gamma+1}. But in your formula, it got  (-1) * (-y)^{gamma+1}.
I think it should be (-1)^{\hat{y}} *(\hat{y}+y*(-1)^{\hat{y}})^{\gamma+1}

**Comment by jhwjhw0123** [comment]
> > > > > It seems that the formula of the 1st-order derivative(the second term) is still wrong.
> > > > 
> > > > 
> > > > You mean the (y+\hat{y}-1)^{\gamma} term?
> > > 
> > > 
> > > I mean the last one, (\hat{y}-y)^{\gamma+1} term
> > 
> > 
> > This part is unlikely to be wrong as if you plug in \gamma = 0 it will recover ordinary cross-entropy derivative. Nevertheless, do you have a result you believe to be correct?
> 
> When \hat{y} is 0, the gradient excluding the log term should be (y)^{\gamma+1}. But in your formula, it got (-1) * (-y)^{\gamma+1}.
> I think it should be (-1)^{\hat{y}} * (\hat{y}+y * (-1)^{\hat{y}})^{\gamma+1}

Yep, it seems you're right. I missed the (-y)^\gamma term when merging the two. It's different from merging ordinary cross-entropy. Thanks for pointing that out.

I'll update the math and script and acknowledge your efforts in improving this.

Best,
Chen
